### PR TITLE
Add support for WaveMatch and WaveMultiPrefix<Op>

### PR DESCRIPTION
**Pull request by jholewinski**
These new DXIL instructions are added to SM 6.5. The valid operations
for <Op> are:

    - BitAnd
    - BitOr
    - BitXor
    - CountBits
    - Product
    - Sum

In HLSL, these are exposed as:

    uint4 WaveMatch(<type> val)
    <type> WaveMultiPrefixBitAnd(<type> val, uint4 mask)
    <type> WaveMultiPrefixBitOr(<type> val, uint4 mask)
    <type> WaveMultiPrefixBitXor(<type> val, uint4 mask)
    uint WaveMultiPrefixCountBits(bool val, uint4 mask)
    <type> WaveMultiPrefixProduct(<type> val, uint4 mask)
    <type> WaveMultiPrefixSum(<type> val, uint4 mask)

In DXIL, these are exposed as:

    [BitAnd,BitOr,BitXor,Product,Sum]
    %dx.types.fouri32 @dx.op.waveMatch.T(i32 %opc, T %val)
    T @dx.op.waveMultiPrefixOp.T(i32 %opc, T %val, i32 %mask_x,
                                 i32 %mask_y, i32 %mask_y, i32 %mask_z,
                                 i8 %operation, i8 %signed)

    [CountBits]
    i32 @dx.op.waveMultiPrefixBitCount(i32 %opc, i1 %val, i32 %mask_x,
                                       i32 %mask_y, i32 %mask_y,
                                       i32 %mask_z)

Scalarization of vector types occur as per the existing wave intrinsics.
For WaveMatch, the match is performed on each scalar and the results
are combined with bitwise AND. For WaveMultiPrefix, the operation is
performed on each scalar and combined into an aggregate.

### CODE REVIEW

File: tools/clang/unittests/HLSL/ExecutionTest.cpp [hunk 1]
```diff
@@ -5770,6 +5799,182 @@ TEST_F(ExecutionTest, WaveIntrinsicsPrefixUintTest) {
...
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixUSum") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs + rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitAnd") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs & rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitOr") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs | rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitXor") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs ^ rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixCountBits") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs + (rhs > 10 ? 1 : 0); };
```
Comment (MrTrillian) [inline tools/clang/unittests/HLSL/ExecutionTest.cpp]: What is this 10? I don't understand how that counts bits.

Comment (jholewinski) [reply tools/clang/unittests/HLSL/ExecutionTest.cpp]: It's a bit of a hack to be able to share the same testing infrastructure as the rest of the MultiPrefix tests. The shader code uses `(data.value > 10)` as the boolean input to `WaveMultiPrefxCountBits`. This reference code does the same by adding 1 to the bit count if the value is greater than 10.

Comment (MrTrillian) [reply tools/clang/unittests/HLSL/ExecutionTest.cpp]: I see `res.x += WaveMultiPrefixCountBits((input.x == 1), mask);`...? In any case, a comment would help :)

**Comment by MrTrillian** [comment]
Nice! Looks like a high quality change to me. I'll leave the final word to @tex3d though because I don't know about what is supposed or not to go into DXIL 1.5.

File: tools/clang/unittests/HLSL/ExecutionTest.cpp (hunk 1 above)
Comment (jholewinski) [reply tools/clang/unittests/HLSL/ExecutionTest.cpp]: Yup, I'll add some comments, and try to clean up this conditional (lots of duplicate bodies).

The shader for this test is in `ShaderOpArithTable.xml`. Specifically: `data.result = WaveMultiPrefixCountBits(data.value > 10, mask);`

The `(input.x == 1)` line is from a compile-only sanity test in `sm_6_5_wave.hlsl`.

**Comment by jholewinski** [comment]
Thanks for the review!

@tex3d, would you prefer I re-submit a new pull request with the fixes squashed into one commit?

**Comment by MrTrillian** [comment]
@jholewinski We can squish as we merge the pull request so there's really no need :)

**Comment by jholewinski** [comment]
Thanks Tex! Are we good to merge now?

**Comment by tex3d** [comment]
> Thanks Tex! Are we good to merge now?

Yes, sorry I missed this comment.  Go ahead and (squash and) merge!

**Comment by jholewinski** [comment]
@tex3d, who is this comment meant for? I don't have write access to the repo to merge the PR.

**Pull request merged**